src/core/border_router/routing_manager.cpp

```diff
@@ -3110,6 +3183,245 @@ void RoutingManager::RsSender::HandleTimer(void)
     return;
 }
 
+#if OPENTHREAD_CONFIG_BORDER_ROUTING_DHCP6_PD_ENABLE
+RoutingManager::PdPrefixManager::PdPrefixManager(Instance &aInstance)
+    : InstanceLocator(aInstance)
+    , mEnabled(false)
+    , mTimer(aInstance)
+{
+    mPrefix.Clear();
+}
+
+Error RoutingManager::PdPrefixManager::GetPrefixInfo(otBorderRoutingPrefixTableEntry &aInfo) const
+{
+    Error     error = kErrorNone;
+    TimeMilli now   = TimerMilli::GetNow();
+
+    VerifyOrExit(IsRunning() && IsValidOmrPrefix(mPrefix), error = kErrorNotFound);
+
+    aInfo.mPrefix = mPrefix;
+    // Compare the time before subtraction in case of the timers have expired recently.
+    aInfo.mPreferredLifetime =
+        (mPreferred && mPreferredUntil > now) ? (mPreferredUntil - now) / TimeMilli::kOneSecondInMsec : 0;
+    aInfo.mValidLifetime = (mValidUntil > now) ? (mValidUntil - now) / TimeMilli::kOneSecondInMsec : 0;
+
+exit:
+    return error;
+}
+
+void RoutingManager::PdPrefixManager::WithdrawPlatformRouterAdvertPrefix(void)
+{
+    Error error = kErrorNone;
+
+    VerifyOrExit(IsValidOmrPrefix(mPrefix), error = kErrorInvalidState);
+
+    mPrefix.SetLength(0);
+    mTimer.Stop();
+
+    Get<RoutingManager>().ScheduleRoutingPolicyEvaluation(kImmediately);
+    LogInfo("Withdrawed platform provided outdated prefix: %s", mPrefix.ToString().AsCString());
+
+exit:
+    if (error != kErrorNone)
+    {
+        LogCrit("Failed to remove platform provided outdated prefix: %s", ErrorToString(error));
+    }
+}
+
+void RoutingManager::PdPrefixManager::DeprecatePlatformRouterAdvertPrefix(void)
+{
+    Error error = kErrorNone;
+
+    VerifyOrExit(IsValidOmrPrefix(mPrefix), error = kErrorInvalidState);
+    mPreferred = false;
+
+    Get<RoutingManager>().ScheduleRoutingPolicyEvaluation(kImmediately);
```

Comment:
When we deprecate the PD prefix, do we want to keep adding it in netdata? (and have it be used by all nodes)
Should we remove it and switch to "generated" local prefix?

Reply:
This logic is inplemented in `OmrPrefixManager::DetermineLocalPrefix`, since we have other cases to consider about and I want to place them in a single place.

- When we receive a PD prefix, we want to remove the generated one from the netdata.
- When we received a different PD prefix, we want to remove the old one from the netdata.

OmrPrefixManager now tracks the published prefix and will remove it when we want to publish another prefix.

To logic is, PD prefix manager will be a source of prefix for OmrPrefixManager, and it does not control the prefix in netdata directly, instead, it will inform OmrPrefixManager to do so when its state changes.

Reply:
Yes, after checking the spec, we should remove it from the network data.

Reply:
SGTM. Thanks.